### PR TITLE
SIMDify upsampling

### PR DESCRIPTION
**Pull request by eustas**
```
+------------+--------+--------+--------+
| cycles     |   2x   |   4x   |   8x   |
+------------+--------+--------+--------+
| Baseline   | 249948 | 377275 | 523975 |
+------------+--------+--------+--------+
| Scalar (1) | 111383 | 162607 | 299005 |
+------------+--------+--------+--------+
| SSE4 (4)   |  48596 |  47257 |  72677 |
+------------+--------+--------+--------+
| AVX2 (8)   |  38341 |  39268 |  45655 |
+------------+--------+--------+--------+
| AVX3 (16)  |  36736 |  44514 |  41769 |
+------------+--------+--------+--------+
```

BEFORE

```
Compr                            Input    Compr            Compr       Compr  Decomp  Butteraugli
Method                          Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
----------------------------------------------------------------------------------------------------------------------------------------------
jxl:fast:d1:resampling=2     136323072  8731417    0.51239555400   1  32.522   2.398  13.27280712   2.98966850267  1.5318928486911312        0
jxl:fast:d1:resampling=4     136323072  2494249    0.14637281648   1  58.210   3.667  54.28284454  11.65082575442  1.7053641800059982        0
jxl:fast:d1:resampling=8     136323072   656189    0.03850787635   1  71.192   5.017 113.40100098  27.19273331255  1.0471344119728105        0
Aggregate:                   136323072  2426710    0.14240933233 ---  51.271   3.533  43.39244180   9.82073974612  1.3985649902105746        0
```

AFTER
```
Compr                            Input    Compr            Compr       Compr  Decomp  Butteraugli
Method                          Pixels     Size              BPP   #    MP/s    MP/s     Distance    Error p norm           BPP*pnorm   Errors
----------------------------------------------------------------------------------------------------------------------------------------------
jxl:fast:d1:resampling=2     136323072  8731417    0.51239555400   1  30.802  20.693  13.27280331   2.98966850189  1.5318928482907228        0
jxl:fast:d1:resampling=4     136323072  2494249    0.14637281648   1  54.076  32.372  54.28282928  11.65082574146  1.7053641781087603        0
jxl:fast:d1:resampling=8     136323072   656189    0.03850787635   1  67.589  50.065 113.40100098  27.19273316916  1.0471344064509187        0
Aggregate:                   136323072  2426710    0.14240933233 ---  48.286  32.248  43.39243358   9.82073972436  1.3985649871117138        0
```

### CODE REVIEW

File: lib/jxl/dec_group.cc [hunk 1]
```diff
@@ -803,12 +803,17 @@ Status DecodeGroup(BitReader* JXL_RESTRICT* JXL_RESTRICT readers,
                       PassesDecoderState::kGroupDataYBorder, dst_rect.xsize(),
                       dst_rect.ysize());
     }
+
+    // TODO(eustas): make "thread-local".
+    auto arena =
```
Comment (veluca93) [inline lib/jxl/dec_group.cc:808]: Can we move this in the dec_state? There are already a few num-thread-dependent things there (IIRC PrepareForThreads is the relevant function, or something similarly named)

File: lib/jxl/dec_reconstruct.cc [hunk 2]
```diff
@@ -612,6 +612,12 @@ Status FinalizeImageRect(
         frame_dim.ysize_padded, 2, 2, &ensure_padding_upsampling_y0,
         &ensure_padding_upsampling_y1);
   }
+  const size_t max_upsampled_xsize = std::max(
+      upsampled_frame_rect.xsize(), upsampled_frame_rect_for_storage.xsize());
+  auto upsampler_arena_storage =
```
Comment (veluca93) [inline lib/jxl/dec_reconstruct.cc:617]: ditto

**Review by veluca93 — APPROVED** [review]

File: lib/jxl/dec_group.cc (hunk 1 above)
Comment (eustas) [reply lib/jxl/dec_group.cc]: Done

File: lib/jxl/dec_reconstruct.cc (hunk 2 above)
Comment (eustas) [reply lib/jxl/dec_reconstruct.cc]: Done.

**Review by veluca93 — CHANGES_REQUESTED** [review]

File: lib/jxl/dec_cache.h [hunk 3]
```diff
@@ -174,6 +175,13 @@ struct PassesDecoderState {
             kApplyImageFeaturesTileDim + 4);
       }
     }
+    size_t max_upsampler_dst_xsize = std::max(
```
[the remark this answers is not in the document]
Comment (eustas) [reply lib/jxl/dec_cache.h:178]: Hmm. Just remembered that ec could have different upsampling factor... Going to update this calculation.

[the remark this answers is not in the document]
Comment (eustas) [reply lib/jxl/dec_cache.h]: Checked the codepaths, +4 is not necessafy.

**Comment by eustas** [comment]
@veluca93 PTAL

**Comment by veluca93** [comment]
> @veluca93 PTAL

LGTM

**Review by veluca93 — APPROVED** [review]

**Pull request merged**